cli/command/disks/commit.go

```diff
@@ -0,0 +1,214 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package disks
+
+import (
+	"github.com/fatih/color"
+	"github.com/opencurve/curveadm/cli/cli"
+	comm "github.com/opencurve/curveadm/internal/common"
+	"github.com/opencurve/curveadm/internal/configure/disks"
+	"github.com/opencurve/curveadm/internal/configure/hosts"
+	"github.com/opencurve/curveadm/internal/errno"
+	"github.com/opencurve/curveadm/internal/storage"
+	tui "github.com/opencurve/curveadm/internal/tui/common"
+	"github.com/opencurve/curveadm/internal/utils"
+	"github.com/spf13/cobra"
+)
+
+const (
+	COMMIT_EXAMPLE = `Examples:
+  $ curveadm disks commit /path/to/disks.yaml  # Commit disks`
+)
+
+type commitOptions struct {
+	filename string
+	slient   bool
+}
+
+func NewCommitCommand(curveadm *cli.CurveAdm) *cobra.Command {
+	var options commitOptions
+
+	cmd := &cobra.Command{
+		Use:     "commit DISKS [OPTIONS]",
+		Short:   "Commit disks",
+		Args:    utils.ExactArgs(1),
+		Example: COMMIT_EXAMPLE,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.filename = args[0]
+			return runCommit(curveadm, options)
+		},
+		DisableFlagsInUseLine: true,
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.slient, "slient", "s", false, "Slient output for disks commit")
+
+	return cmd
+}
+
+func readAndCheckDisks(curveadm *cli.CurveAdm, options commitOptions) (string, error) {
+	// 1) read disks from file
+	if !utils.PathExist(options.filename) {
+		return "", errno.ERR_DISKS_FILE_NOT_FOUND.
+			F("%s: no such file", utils.AbsPath(options.filename))
+	}
+	data, err := utils.ReadFile(options.filename)
+	if err != nil {
+		return data, errno.ERR_READ_DISKS_FILE_FAILED.E(err)
+	}
+
+	// 2) display difference
+	oldData := curveadm.Disks()
+	if !options.slient {
+		diff := utils.Diff(oldData, data)
+		curveadm.WriteOutln(diff)
+	}
+
+	// 3) check disks data
+	_, err = disks.ParseDisks(data)
+	return data, err
+}
+
+func updateDisk(data string, curveadm *cli.CurveAdm) error {
+	var err error
+	// hosts_data := curveadm.Hosts()
+	hcs, err := hosts.ParseHosts(curveadm.Hosts())
+	if err != nil {
+		return err
+	}
+	dcs, err := disks.ParseDisks(data)
+	if err != nil {
+		return err
+	}
+
+	var toRecord bool
+	diskMapList := []map[string][]string{}
+	// write disk records
+	for _, hc := range hcs {
+		diskMap := make(map[string][]string)
+		for _, dc := range dcs {
```

Comment:
You can wrap the below logic into one function.
BTW: i think you can make the function `ParseDisk` more powerful, and let it return the final disk config which already handle `hosts_exclude` and `hosts_exclude`, and then insert the records into database.